OkvisSLAMSystem.cpp

```diff
@@ -286,20 +310,190 @@ namespace ark {
+                                            correspondences, inliers, transformEstimate);
+
+        //std::cout << transformEstimate.matrix() << std::endl;
+
+        return true;
+    }
+
+
+    std::shared_ptr<SparseMap<DBoW2::FBRISK::TDescriptor, DBoW2::FBRISK>> OkvisSLAMSystem::mergeMaps(
+            std::shared_ptr<SparseMap<DBoW2::FBRISK::TDescriptor, DBoW2::FBRISK>> olderMap,
+            std::shared_ptr<SparseMap<DBoW2::FBRISK::TDescriptor, DBoW2::FBRISK>> currentMap, MapKeyFrame::Ptr kf,
+            MapKeyFrame::Ptr loop_kf, Eigen::Affine3d &transformEstimate) {
+    
+        // Loop is detected. Merge smaller map into bigger map and save correction
+        std::shared_ptr<SparseMap<DBoW2::FBRISK::TDescriptor, DBoW2::FBRISK>> mapA;
+        std::shared_ptr<SparseMap<DBoW2::FBRISK::TDescriptor, DBoW2::FBRISK>> mapB;
+        Eigen::Matrix4d correction;
+        Eigen::Matrix4d kfCorrection;
+        Eigen::Matrix4d T_KfKloop = kf->T_SC_[2]*transformEstimate.inverse().matrix()*kf->T_SC_[2].inverse();
+        if (false && olderMap->frameMap_.size() > currentMap->frameMap_.size()) {
+            //TODO - this case is a work in progress - not yet compatible with SlamReplaying.cpp and SlamDemo435i.cpp
+            //merge current map into older map
+            mapA = currentMap;
+            mapB = olderMap;
+            correction = T_KfKloop * loop_kf->T_WS() * kf->T_WS().inverse();
+            kfCorrection = correction;
+            correction_ = correction * correction_;
+            
+        } else {
+            //merge older map into current map
+            mapA = olderMap;
+            mapB = currentMap;
+            correction = kf->T_WS() * (T_KfKloop * loop_kf->T_WS()).inverse();
+            kfCorrection = Eigen::Matrix4d::Identity();
+        }
+
+        //adding keyframes from mapA to mapB
+        for (auto framePair = mapA->frameMap_.begin(); framePair != mapA->frameMap_.end(); framePair++) {
+            auto frame = framePair->second;
+            frame->setOptimizedTransform(correction * frame->T_WS());
+            mapB->frameMap_[frame->frameId_] = frame;
+		}
+
+        //adding constraints and poses to mapB's pose graph
+        mapB->graph_.constraintMutex.lock();
+        mapA->graph_.constraintMutex.lock();
+        mapB->graph_.constraints_.insert(
+            mapB->graph_.constraints_.end(),
+            mapA->graph_.constraints_.begin(),
+            mapA->graph_.constraints_.end()
+        );
+        mapA->graph_.constraintMutex.unlock();
+        for (auto framePair = mapA->frameMap_.begin(); framePair != mapA->frameMap_.end(); framePair++) {
+            mapB->graph_.poses_.insert(std::pair<int,GraphPose>(
+                framePair->first,framePair->second->T_WS()));
+		}
+        mapB->graph_.constraintMutex.unlock();
+
+        //adding current keyframe to mapB
+        kf->T_WS_ = kfCorrection * kf->T_WS_;
```

Comment:
From line 475-494, this code seems for the most part C+P'd out of SparseMap.h's addKeyFrame(). Can you call that function instead of pulling the implementation here, or write a new function if something else is needed? Encapsulation